### PR TITLE
fix(recovery): send u2f counter in params

### LINKED ISSUE #2933: U2F counter restoration not working
When restoring trezor one (using trezor suite), the u2f counter is reset to 0. That can make authentication fail if counter is lower than one used before. Should be unix time ?

trezor used : trezor one
firmware : 1.9.3
trezor suite : 20.10.1
website used for checking counter, with chrome on windows 10 : https://mdp.github.io/u2fdemo/

I have tested with wallet.trezor.io, it is working, counter is restored with unixtime. This is an issue with trezor suite

### PR DESCRIPTION
**Pull request by mroz22**
todo: 
- [x] update types in trezor-connect(PR https://github.com/trezor/connect/pull/710)
- [x] wait for trezor-connect@8.1.19
- [x] not sure about if I can use capabilities `Capability_U2F`. This feature seems to be supported long time before capabilites have been introduced https://github.com/satoshilabs/mytrezor/blob/755b2e27028dcefbc79a0f5032672e417492ceed/app/scripts/config.js#L33 ?
___
will fix #2933

### CODE REVIEW

**Comment by mroz22** [comment]
It seems that I can't use capabilities for this. This feature seems to be supported long time before capabilites have been introduced https://github.com/satoshilabs/mytrezor/blob/755b2e27028dcefbc79a0f5032672e417492ceed/app/scripts/config.js#L33

File: packages/suite/src/actions/recovery/recoveryActions.ts [hunk 1]
```diff
@@ -79,14 +81,23 @@ const recoverDevice = () => async (dispatch: Dispatch, getState: GetState) => {
...
         type: advancedRecovery ? 1 : 0,
         word_count: wordsCount,
         passphrase_protection: DEVICE.DEFAULT_PASSPHRASE_PROTECTION,
+        // @ts-ignore https://github.com/trezor/connect/pull/710
+        enforce_wordlist: true,
         device: {
             path: device.path,
         },
-    });
+    } as const;
```
Comment (keraf) [inline packages/suite/src/actions/recovery/recoveryActions.ts]: const as const? Why is that needed?

File: packages/suite/src/actions/recovery/recoveryActions.ts [hunk 2]
```diff
@@ -79,14 +81,23 @@ const recoverDevice = () => async (dispatch: Dispatch, getState: GetState) => {
     dispatch(setError(''));
     dispatch(setStatus('in-progress'));
 
-    const response = await TrezorConnect.recoveryDevice({
+    const params = {
         type: advancedRecovery ? 1 : 0,
         word_count: wordsCount,
         passphrase_protection: DEVICE.DEFAULT_PASSPHRASE_PROTECTION,
+        // @ts-ignore https://github.com/trezor/connect/pull/710
```
Comment (keraf) [inline packages/suite/src/actions/recovery/recoveryActions.ts]: Looks like this is going to make it in connect 8.1.19, the ignore can be removed once rebased with Szymon's update.

File: packages/suite/src/actions/recovery/recoveryActions.ts [hunk 3]
```diff
@@ -79,14 +81,23 @@ const recoverDevice = () => async (dispatch: Dispatch, getState: GetState) => {
...
+        enforce_wordlist: true,
         device: {
             path: device.path,
         },
-    });
+    } as const;
+
+    if (device.features.capabilities?.includes('Capability_U2F')) {
+        // @ts-ignore todo: update connect types
+        params.u2f_counter = Math.floor(Date.now() / 1000);
```
Comment (keraf) [inline packages/suite/src/actions/recovery/recoveryActions.ts:97]: Same comment as above (regarding the merged PR and connect update).

**Review by keraf — CHANGES_REQUESTED** [review]

File: packages/suite/src/actions/recovery/recoveryActions.ts (hunk 2 above)
Comment (mroz22) [reply packages/suite/src/actions/recovery/recoveryActions.ts]: https://www.npmjs.com/package/trezor-connect waiting for 8.1.19 to be releaed

File: packages/suite/src/actions/recovery/recoveryActions.ts (hunk 1 above)
Comment (mroz22) [reply packages/suite/src/actions/recovery/recoveryActions.ts]: type and word_count are number literals. https://github.com/trezor/connect/blob/b5dea6e305421586c1e6ae13b653331facddbd2a/src/ts/types/trezor/management.d.ts#L46
If you have any idea how to do this in a more elegant way I am all ears

Comment (mroz22) [reply packages/suite/src/actions/recovery/recoveryActions.ts]: I did it in another way.

File: packages/suite/src/actions/recovery/recoveryActions.ts [hunk 4]
```diff
@@ -85,14 +86,21 @@ const recoverDevice = () => async (dispatch: Dispatch, getState: GetState) => {
     dispatch(setError(''));
     dispatch(setStatus('in-progress'));
 
-    const response = await TrezorConnect.recoveryDevice({
+    const params: Parameters<typeof TrezorConnect['recoveryDevice']>[0] = {
```
Comment (keraf) [inline packages/suite/src/actions/recovery/recoveryActions.ts]: Looks like a weird way for defining a type. Isn't there an existing type ready to use for this?

Comment (tomexx) [reply packages/suite/src/actions/recovery/recoveryActions.ts]: agree with Raf

**Review by tomexx — CHANGES_REQUESTED** [review]

File: packages/suite/src/actions/recovery/recoveryActions.ts (hunk 4 above)
Comment (mroz22) [reply packages/suite/src/actions/recovery/recoveryActions.ts]: Well, I am not defining a type here. I am reusing an existing type imported from trezor-connect. If you have any better idea how to do it, please let me know.

Comment (mroz22) [reply packages/suite/src/actions/recovery/recoveryActions.ts]: Of course I could do something like this, but it feels much less readable to me

```
 const response = await TrezorConnect.recoveryDevice({
        type: advancedRecovery ? 1 : 0,
        word_count: wordsCount,
        passphrase_protection: DEVICE.DEFAULT_PASSPHRASE_PROTECTION,
        enforce_wordlist: true,
        device: {
            path: device.path,
        },
        u2f_counter: device.features.capabilities?.includes('Capability_U2F')
            ? Math.floor(Date.now() / 1000)
            : undefined,
    });
```

Comment (mroz22) [reply packages/suite/src/actions/recovery/recoveryActions.ts]: # :exploding_head: :gun:

Comment (szymonlesisz) [reply packages/suite/src/actions/recovery/recoveryActions.ts]: import { RecoveryDevice } from 'trezor-connect';

:)

**Review by szymonlesisz — CHANGES_REQUESTED** [review]

**Comment by szymonlesisz** [comment]
and to your concern about capability. connect will use default capabilities for the device with older FW (which doesn't have `features.capabilites` yet)
https://github.com/trezor/connect/blob/develop/src/js/utils/deviceFeaturesUtils.js#L27-L28

and `Capability_U2F` is there (14), so it's always there

**Comment by mroz22** [comment]
Thanks @szymonlesisz I didn't know this was also exported. Or maybe I forgot

**Review by tomexx — APPROVED** [review]

**Review by szymonlesisz — APPROVED** [review]

**Pull request merged**